app/uk/gov/hmrc/gform/formtemplate/JsonSchemeValidator.scala

```diff
@@ -67,59 +43,193 @@ object JsonSchemeValidator {
-                deduplicatedConditionalValidationErrors.map { case (errorProperty, errorLocation) =>
-                  val deduplicatedConditionalRequirements: List[String] = {
-                    val distinctProperties = conditionalRequirements(errorProperty).map(_.getRequiredProperty).distinct
-                    val groupedProperties = conditionalRequirements(errorProperty).groupBy(_.getRequiredProperty)
-                    distinctProperties.map(propertyKey => propertyKey -> groupedProperties(propertyKey)).map {
-                      case (property, requiredValues) =>
-                        s"$property: [${requiredValues.map(_.getRequiredValue).mkString(", ")}]"
-                    }
-                  }
-
-                  s"$errorLocation: Property $errorProperty can only be used with ${deduplicatedConditionalRequirements.mkString(", ")}"
-                }
-
-              val typeErrors =
-                errors.toList.slice(indexOfAnyOf, errors.length).filter(_.keyword == "type").map(_.getMessage)
-
-              // Using unsafe because errors is an NEL and indexOfAnyOf is > 0 in this branch
-              NonEmptyList.fromListUnsafe(
-                errors
-                  .map(_.getMessage)
-                  .toList
-                  .slice(0, indexOfAnyOf) ++ typeErrors ++ formattedConditionalValidationErrors
-              )
-            } else {
-              errors.map(_.getMessage)
+          val fullError: NonEmptyList[String] = {
+            val errorsWithParsedTypeErrors = parseTypeErrors(errors, json)
+
+            maybeFirstSchemaValidationErrorIndex(errors) match {
+              case None => errorsWithParsedTypeErrors.map(_.getMessage)
+              case Some(firstDependencyIndex) =>
+                val conditionalValidationErrors: List[String] =
+                  parseConditionalValidationErrors(errors, firstDependencyIndex, schema, json)
+
+                constructFullError(errorsWithParsedTypeErrors, conditionalValidationErrors, firstDependencyIndex)
             }
           }
 
-          SchemaValidationException(conditionalValidationErrorMessages)
+          SchemaValidationException(fullError)
         }.toEither
     }
 
+  private def parseTypeErrors(
+    errors: NonEmptyList[ValidationError],
+    json: Json
+  ): NonEmptyList[ValidationError] =
+    errors.map { error =>
+      if (error.keyword == "type") {
+        val remainingSections = error.location.split("/").toList.tail
+        val errorProperty = remainingSections.last
+
+        val errorLocationId: String = getErrorLocationId(json, remainingSections.dropRight(1)) match {
+          case None => remainingSections.mkString("/")
+          case Some(value) =>
+            value.asString match {
+              case None                                => remainingSections.mkString("/")
+              case Some(errorLocationIdString: String) => errorLocationIdString
+            }
+        }
```

Comment:
After little reshuffling of operations we can see some repetition. See the comment in the code suggestion

```suggestion
        val locationSplitted = error.location.split("/").toList
        val remainingSections = locationSplitted.dropRight(1).tail
        val errorProperty = locationSplitted.last

        // This now looks exactly like bit from `tryConvertErrorLocationToId` - so candidate for extracting to new method
        val errorLocationId: String = getErrorLocationId(json, remainingSections)
          .flatMap(_.asString)
          .getOrElse(error.location)
```